<a href="https://colab.research.google.com/github/Ankita-chouhan-github/Book-Recommendation-System/blob/main/Unsupervised_ML_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Unsupervised ML Capstone Project: Book Recommemdation System**

The objective of a book recommendation system is to predict buyer’s interest and recommend books to them respectively. That system can take into regard many parameters like book content and quality by filtering user reviews.

In [3]:
import re
import pickle
import operator
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from collections import Counter
from scipy.sparse import csr_matrix
from pandas.api.types import is_numeric_dtype
from sklearn.neighbors import NearestNeighbors
from sklearn.feature_extraction import DictVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

import warnings
warnings.filterwarnings("ignore")

In [12]:
books = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Unsupervised ML Capstone Project/Books.csv',low_memory=False,error_bad_lines=False,encoding="latin-1")
ratings = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Unsupervised ML Capstone Project/Ratings.csv',low_memory=False,error_bad_lines=False,encoding="latin-1")
users = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Unsupervised ML Capstone Project/Users.csv',low_memory=False,error_bad_lines=False,encoding="latin-1")

print("Books Data:    ", books.shape)
print("Users Data:    ", users.shape)
print("Books-ratings: ", ratings.shape)

Books Data:     (271360, 8)
Users Data:     (278858, 3)
Books-ratings:  (1149780, 3)


#**Books Data**


---
**Data Description**

Books are identified by their respective ISBN. Invalid ISBNs have already been removed from the dataset. Moreover, some content-based information is given (Book-Title, Book-Author, Year-Of-Publication, Publisher), obtained from Amazon Web Services. Note that in the case of several authors, only the first is provided. URLs linking to cover images are also given, appearing in three different flavors (Image-URL-S, Image-URL-M, Image-URL-L), i.e., small, medium, large. These URLs point to the Amazon website.



In [4]:
print("Columns: ", list(books.columns))
books.head()

Columns:  ['ISBN,Book-Title,Book-Author,Year-Of-Publication,Publisher,Image-URL-S,Image-URL-M,Image-URL-L']


ISBN,Book-Title,Book-Author,Year-Of-Publication,Publisher,Image-URL-S,Image-URL-M,Image-URL-L
0  0195153448,Classical Mythology,Mark P. O. Morf...                                           
1  0002005018,Clara Callan,Richard Bruce Wright,2...                                           
2  0060973129,Decision in Normandy,Carlo D'Este,1...                                           
3  0374157065,Flu: The Story of the Great Influen...                                           
4  0399135782,The Kitchen God's Wife,Amy Tan,1991...

In [5]:
print(books.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249936 entries, 0 to 249935
Data columns (total 1 columns):
 #   Column                                                                                         Non-Null Count   Dtype 
---  ------                                                                                         --------------   ----- 
 0   ISBN,Book-Title,Book-Author,Year-Of-Publication,Publisher,Image-URL-S,Image-URL-M,Image-URL-L  249936 non-null  object
dtypes: object(1)
memory usage: 1.9+ MB
None


In [6]:
books.describe()

ISBN,Book-Title,Book-Author,Year-Of-Publication,Publisher,Image-URL-S,Image-URL-M,Image-URL-L
count                                              249936                                           
unique                                             249936                                           
top     0195153448,Classical Mythology,Mark P. O. Morf...                                           
freq                                                    1

In [7]:
# Missing value percentage
print(books.isnull().sum()/len(books)*100)

ISBN,Book-Title,Book-Author,Year-Of-Publication,Publisher,Image-URL-S,Image-URL-M,Image-URL-L    0.0
dtype: float64


In [13]:
# Checking  for  null value in book author
books[books['Book-Author'].isna()]

ISBN                                         Book-Title  \
187689  9627982032  The Credit Suisse Guide to Managing Your Perso...   

       Book-Author Year-Of-Publication                       Publisher  \
187689         NaN                1995  Edinburgh Financial Publishing   

                                              Image-URL-S  \
187689  http://images.amazon.com/images/P/9627982032.0...   

                                              Image-URL-M  \
187689  http://images.amazon.com/images/P/9627982032.0...   

                                              Image-URL-L  
187689  http://images.amazon.com/images/P/9627982032.0...

In [15]:
#Filling the null value
books.loc[187689, 'Book-Author'] = 'Larissa Anne Downes'

In [16]:
# Checking null values in publisher
books[books['Publisher'].isna()]

ISBN       Book-Title      Book-Author Year-Of-Publication  \
128890  193169656X      Tyrant Moon  Elaine Corvidae                2002   
129037  1931696993  Finders Keepers  Linnea Sinclair                2001   

       Publisher                                        Image-URL-S  \
128890       NaN  http://images.amazon.com/images/P/193169656X.0...   
129037       NaN  http://images.amazon.com/images/P/1931696993.0...   

                                              Image-URL-M  \
128890  http://images.amazon.com/images/P/193169656X.0...   
129037  http://images.amazon.com/images/P/1931696993.0...   

                                              Image-URL-L  
128890  http://images.amazon.com/images/P/193169656X.0...  
129037  http://images.amazon.com/images/P/1931696993.0...

In [17]:
#Replacing NaNs with correct  values
books.loc[128890, 'Publisher'] = 'Mundania Press LLC'
books.loc[129037, 'Publisher'] = 'Bantam'

In [18]:
#insepcting the values in year of publication
books['Year-Of-Publication'].unique()

array(['2002', '2001', '1991', '1999', '2000', '1993', '1996', '1988',
       '2004', '1998', '1994', '2003', '1997', '1983', '1979', '1995',
       '1982', '1985', '1992', '1986', '1978', '1980', '1952', '1987',
       '1990', '1981', '1989', '1984', '0', '1968', '1961', '1958',
       '1974', '1976', '1971', '1977', '1975', '1965', '1941', '1970',
       '1962', '1973', '1972', '1960', '1966', '1920', '1956', '1959',
       '1953', '1951', '1942', '1963', '1964', '1969', '1954', '1950',
       '1967', '2005', '1957', '1940', '1937', '1955', '1946', '1936',
       '1930', '2011', '1925', '1948', '1943', '1947', '1945', '1923',
       '2020', '1939', '1926', '1938', '2030', '1911', '1904', '1949',
       '1932', '1928', '1929', '1927', '1931', '1914', '2050', '1934',
       '1910', '1933', '1902', '1924', '1921', '1900', '2038', '2026',
       '1944', '1917', '1901', '2010', '1908', '1906', '1935', '1806',
       '2021', '2012', '2006', 'DK Publishing Inc', 'Gallimard', '1909',
       

Name of few publication companies have been incorrectly put in this column.There are values such as 0 ,2024,2030 etc. which is also not possible .Let's rectify these mistakes

In [19]:
# correcting this error
books[books['Year-Of-Publication'] == 'DK Publishing Inc']

ISBN                                         Book-Title  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

       Book-Author Year-Of-Publication  \
209538        2000   DK Publishing Inc   
221678        2000   DK Publishing Inc   

                                                Publisher  \
209538  http://images.amazon.com/images/P/078946697X.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-S  \
209538  http://images.amazon.com/images/P/078946697X.0...   
221678  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-M Image-URL-L  
209538  http://images.amazon.com/images/P/078946697X.0...         NaN  
221678  http://images.amazon.com/images/P/0789466953.0...         NaN

In [20]:
# on searching for these  books we came to know about its authors
#ISBN '078946697X'
books.loc[books.ISBN == '078946697X','Year-Of-Publication'] = 2000
books.loc[books.ISBN == '078946697X','Book-Author'] = "Michael Teitelbaum"
books.loc[books.ISBN == '078946697X','Publisher'] = "DK Publishing Inc"
books.loc[books.ISBN == '078946697X','Book-Title'] = "DK Readers: Creating the X-Men, How It All Began (Level 4: Proficient Readers)"

#ISBN '0789466953'
books.loc[books.ISBN == '0789466953','Year-Of-Publication'] = 2000
books.loc[books.ISBN == '0789466953','Book-Author'] = "James Buckley"
books.loc[books.ISBN == '0789466953','Publisher'] = "DK Publishing Inc"
books.loc[books.ISBN == '0789466953','Book-Title'] = "DK Readers: Creating the X-Men, How Comic Books Come to Life (Level 4: Proficient Readers)"


In [21]:
#checking the rows having 'Gallimard' as yearOfPublication
books.loc[books['Year-Of-Publication'] == 'Gallimard',:]

ISBN                                         Book-Title  \
220731  2070426769  Peuple du ciel, suivi de 'Les Bergers\";Jean-M...   

       Book-Author Year-Of-Publication  \
220731        2003           Gallimard   

                                                Publisher  \
220731  http://images.amazon.com/images/P/2070426769.0...   

                                              Image-URL-S  \
220731  http://images.amazon.com/images/P/2070426769.0...   

                                              Image-URL-M Image-URL-L  
220731  http://images.amazon.com/images/P/2070426769.0...         NaN

In [22]:
books.loc[books.ISBN=='2070426769','Year-Of-Publication']=2003
books.loc[books.ISBN=='2070426769','Book-Author']='Jean-Marie Gustave Le Cl�©zio'
books.loc[books.ISBN=='2070426769','Publisher']='Gallimard'
books.loc[books.ISBN=='2070426769','Book-Title']="Peuple du ciel, suivi de 'Les Bergers"


In [23]:
#Checking if the corrections are in place
books.loc[books['ISBN'].isin(['2070426769','0789466953','078946697X'])]

ISBN                                         Book-Title  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
220731  2070426769              Peuple du ciel, suivi de 'Les Bergers   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

                            Book-Author Year-Of-Publication  \
209538               Michael Teitelbaum                2000   
220731  Jean-Marie Gustave Le Cl�©zio                2003   
221678                    James Buckley                2000   

                Publisher                                        Image-URL-S  \
209538  DK Publishing Inc  http://images.amazon.com/images/P/078946697X.0...   
220731          Gallimard  http://images.amazon.com/images/P/2070426769.0...   
221678  DK Publishing Inc  http://images.amazon.com/images/P/0789466953.0...   

                                              Image-URL-M Image-URL-L  
209538  http://images.amazon.com/images/P/078946697X.0...         NaN  
220731  http://images.amazon.com/images/P/2070426769.0...         NaN  
221678  http://images.amazon.com/images/P/0789466953.0...         NaN

In [24]:
# changing dtype of year of publication
books['Year-Of-Publication'] =books['Year-Of-Publication'].astype(int)

In [25]:
# something is off about years of publication like:
books[(books['Year-Of-Publication'] > 0) & (books['Year-Of-Publication'] < 1800)]

ISBN                        Book-Title       Book-Author  \
227531  9643112136  Dalan-i bihisht (Dastan-i Irani)       Nazi Safavi   
253750  964442011X                      Tasht-i khun  Isma°il Fasih   

        Year-Of-Publication            Publisher  \
227531                 1378  Intisharat-i Quqnus   
253750                 1376       Nashr-i Alburz   

                                              Image-URL-S  \
227531  http://images.amazon.com/images/P/9643112136.0...   
253750  http://images.amazon.com/images/P/964442011X.0...   

                                              Image-URL-M  \
227531  http://images.amazon.com/images/P/9643112136.0...   
253750  http://images.amazon.com/images/P/964442011X.0...   

                                              Image-URL-L  
227531  http://images.amazon.com/images/P/9643112136.0...  
253750  http://images.amazon.com/images/P/964442011X.0...

In [26]:
#replacing with correct  values
books.loc[books.ISBN=='	9643112136','Year-Of-Publication'] = 2010
books.loc[books.ISBN=='964442011X', 'Year-Of-Publication'] = 1991

In [27]:
#Sustituting np.Nan in rows with year=0 or  greater than the current year,2022.
books.loc[(books['Year-Of-Publication'] > 2022) | (books['Year-Of-Publication'] == 0),'Year-Of-Publication'] = np.NAN

# replacing NaN values with median value of Year-Of-Publication
books['Year-Of-Publication'].fillna(int(books['Year-Of-Publication'].median()), inplace=True)

In [28]:
books['Book-Author'].value_counts()

Agatha Christie         632
William Shakespeare     567
Stephen King            524
Ann M. Martin           423
Carolyn Keene           373
                       ... 
R. Gelman                 1
Ellery R. Sheets          1
Artiste C. Arthur         1
Emily Pearl Kingsley      1
Christopher  Biffle       1
Name: Book-Author, Length: 102023, dtype: int64

In [29]:
books['Publisher'].value_counts()

Harlequin                        7535
Silhouette                       4220
Pocket                           3905
Ballantine Books                 3783
Bantam Books                     3646
                                 ... 
Windstream Publishing Company       1
Polaris Books                       1
Hannover House                      1
Amber Quill Press, LLC.             1
Connaught                           1
Name: Publisher, Length: 16804, dtype: int64

In [30]:
#Inspecting the missing values in 'Publisher' column
books.loc[books.Publisher.isnull(),:]

Empty DataFrame
Columns: [ISBN, Book-Title, Book-Author, Year-Of-Publication, Publisher, Image-URL-S, Image-URL-M, Image-URL-L]
Index: []

In [32]:
#Uppercasing the ISBN numbers
books['ISBN']=books['ISBN'].str.upper()

In [33]:
books[books['Book-Title']=='Emma']

ISBN Book-Title       Book-Author  Year-Of-Publication  \
1495    0451523067       Emma       Jane Austen               1994.0   
8763    055321019X       Emma       Jane Austen               1981.0   
11571   0553212737       Emma       JANE AUSTEN               1984.0   
14263   0140258175       Emma       Jane Austen               1996.0   
45955   B000051WXP       Emma       Jane Austen               1996.0   
53513   0760701652       Emma       Jane Austen               1996.0   
90473   2253140732       Emma       Jane Austen               1997.0   
106616  0460010247       Emma       Jane Austen               1976.0   
133269  0553211595       Emma       Jane Austen               1984.0   
166404  0681218894       Emma       Jane Austen               1996.0   
166534  0749336269       Emma       Jane Austen               1996.0   
171359  0553234315       Emma  Charlotte Bronte               1983.0   
174651  0749336412       Emma       Jane Austen               1996.0   
212084  0786881836       Emma       Jane Austen               1996.0   
261368  0760701660       Emma       Jane Austen               1996.0   

                        Publisher  \
1495    New Amer Library Classics   
8763                 Bantam Books   
11571                      Bantam   
14263        Penguin Putnam~trade   
45955                Random House   
53513          Barnes Noble Books   
90473                         LGF   
106616    J m Dent &amp; Sons Ltd   
133269                     Bantam   
166404      Hunter Publishing+inc   
166534                   Mandarin   
171359          Bantam Books (Mm)   
174651                      Arrow   
212084             Hyperion Books   
261368    Barnes &amp; Noble Inc.   

                                              Image-URL-S  \
1495    http://images.amazon.com/images/P/0451523067.0...   
8763    http://images.amazon.com/images/P/055321019X.0...   
11571   http://images.amazon.com/images/P/0553212737.0...   
14263   http://images.amazon.com/images/P/0140258175.0...   
45955   http://images.amazon.com/images/P/B000051WXP.0...   
53513   http://images.amazon.com/images/P/0760701652.0...   
90473   http://images.amazon.com/images/P/2253140732.0...   
106616  http://images.amazon.com/images/P/0460010247.0...   
133269  http://images.amazon.com/images/P/0553211595.0...   
166404  http://images.amazon.com/images/P/0681218894.0...   
166534  http://images.amazon.com/images/P/0749336269.0...   
171359  http://images.amazon.com/images/P/0553234315.0...   
174651  http://images.amazon.com/images/P/0749336412.0...   
212084  http://images.amazon.com/images/P/0786881836.0...   
261368  http://images.amazon.com/images/P/0760701660.0...   

                                              Image-URL-M  \
1495    http://images.amazon.com/images/P/0451523067.0...   
8763    http://images.amazon.com/images/P/055321019X.0...   
11571   http://images.amazon.com/images/P/0553212737.0...   
14263   http://images.amazon.com/images/P/0140258175.0...   
45955   http://images.amazon.com/images/P/B000051WXP.0...   
53513   http://images.amazon.com/images/P/0760701652.0...   
90473   http://images.amazon.com/images/P/2253140732.0...   
106616  http://images.amazon.com/images/P/0460010247.0...   
133269  http://images.amazon.com/images/P/0553211595.0...   
166404  http://images.amazon.com/images/P/0681218894.0...   
166534  http://images.amazon.com/images/P/0749336269.0...   
171359  http://images.amazon.com/images/P/0553234315.0...   
174651  http://images.amazon.com/images/P/0749336412.0...   
212084  http://images.amazon.com/images/P/0786881836.0...   
261368  http://images.amazon.com/images/P/0760701660.0...   

                                              Image-URL-L  
1495    http://images.amazon.com/images/P/0451523067.0...  
8763    http://images.amazon.com/images/P/055321019X.0...  
11571   http://images.amazon.com/images/P/0553212737.0...  
14263   http://images.amazon.com/images/P/0140258175.0...  
45955   http:/

ISBN numbers for books with the same title and author would change if the year of release (revised editions),regional versions or publishing companies are different. So, let's look for rows where entire column values are identical.

In [34]:
#Checking for duplicates in books_df
books[books.duplicated()]

ISBN                                         Book-Title  \
6635    002542730X  Politically Correct Bedtime Stories: Modern Ta...   
11925   014062080X      The Scarlet Letter (Penguin Popular Classics)   
16298   097089726X                                             Jasper   
19266   039592720X                            Interpreter of Maladies   
22339   042512164X                                     Capital Crimes   
...            ...                                                ...   
267020  088404632X     Dianetics: The Modern Science of Mental Health   
269297  055215038X                                        The Sleeper   
269552  037376099X  Patient Nurse  (Man Of The Month) (Harlequin D...   
270285  189481505X                                                I-O   
270483  039304632X                          The Voyage of the Narwhal   

                Book-Author  Year-Of-Publication                  Publisher  \
6635      James Finn Garner               1994.0  John Wiley &amp; Sons Inc   
11925   Nathaniel Hawthorne               1994.0          Penguin Books Ltd   
16298        Michelle Groce               2003.0     Novello Festival Press   
19266         Jhumpa Lahiri               1999.0        Houghton Mifflin Co   
22339      Lawrence Sanders               2000.0   Berkley Publishing Group   
...                     ...                  ...                        ...   
267020       L. Ron Hubbard               1995.0   Bridge Publications (CA)   
269297        Gillian White               2003.0                Corgi Books   
269552         Diana Palmer               1997.0                 Silhouette   
270285          Simon Logan               2002.0                 Prime (OH)   
270483       Andrea Barrett               1998.0    W W Norton &amp; Co Inc   

                                              Image-URL-S  \
6635    http://images.amazon.com/images/P/002542730X.0...   
11925   http://images.amazon.com/images/P/014062080X.0...   
16298   http://images.amazon.com/images/P/097089726X.0...   
19266   http://images.amazon.com/images/P/039592720X.0...   
22339   http://images.amazon.com/images/P/042512164X.0...   
...                                                   ...   
267020  http://images.amazon.com/images/P/088404632X.0...   
269297  http://images.amazon.com/images/P/055215038X.0...   
269552  http://images.amazon.com/images/P/037376099X.0...   
270285  http://images.amazon.com/images/P/189481505X.0...   
270483  http://images.amazon.com/images/P/039304632X.0...   

                                              Image-URL-M  \
6635    http://images.amazon.com/images/P/002542730X.0...   
11925   http://images.amazon.com/images/P/014062080X.0...   
16298   http://images.amazon.com/images/P/097089726X.0...   
19266   http://images.amazon.com/images/P/039592720X.0...   
22339   http://images.amazon.com/images/P/042512164X.0...   
...                                                   ...   
267020  http://images.amazon.com/images/P/088404632X.0...   
269297  http://images.amazon.com/images/P/055215038X.0...   
269552  http://images.amazon.com/images/P/037376099X.0...   
270285  http://images.amazon.com/images/P/189481505X.0...   
270483  http://images.amazon.com/images/P/039304632X.0...   

                                              Image-URL-L  
6635    http://images.amazon.com/images/P/002542730X.0...  
11925   http://images.amazon.com/images/P/014062080X.0...  
16298   http://images.amazon.com/images/P/097089726X.0...  
19266   http://images.amazon.com/images/P/039592720X.0...  
22339   http://images.amazon.com/images/P/042512164X.0...  
...                                                   ...  
267020  http://images.amazon.com/images/P/088404632X.0...  
269297  http://images.amazon.com/images/P/055215038X.0...  
269552  http://images.amazon.com/images/P/037376099X.0...  
270285  http://images.amazon.com/images/P/189481505X.0...  
270483  http://images.amazon.com/images/P/039304632X.0...  

[

In [35]:
#Dropping the rows with the entire column values are duplicated
books.drop_duplicates(keep="first",inplace=True)
books.reset_index(drop=True,inplace=True)

In [36]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271047 entries, 0 to 271046
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ISBN                 271047 non-null  object 
 1   Book-Title           271047 non-null  object 
 2   Book-Author          271047 non-null  object 
 3   Year-Of-Publication  271047 non-null  float64
 4   Publisher            271047 non-null  object 
 5   Image-URL-S          271047 non-null  object 
 6   Image-URL-M          271047 non-null  object 
 7   Image-URL-L          271044 non-null  object 
dtypes: float64(1), object(7)
memory usage: 16.5+ MB


#**Users Data**


---
**Data Description**

Contains the users. Note that user IDs (User-ID) have been anonymized and map to integers. Demographic data is provided (Location, Age) if available. Otherwise, these fields contain NULL values.



In [37]:
#displaying the top 10 and bottom 10 rows of the dataframe
pd.concat([users.head(10),users.tail(10)],axis=0)

User-ID                            Location   Age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
5             6       santa monica, california, usa  61.0
6             7                 washington, dc, usa   NaN
7             8            timmins, ontario, canada   NaN
8             9          germantown, tennessee, usa   NaN
9            10          albacete, wisconsin, spain  26.0
278848   278849         georgetown, ontario, canada  23.0
278849   278850          sergnano, lombardia, italy   NaN
278850   278851                  dallas, texas, usa  33.0
278851   278852     brisbane, queensland, australia  32.0
278852   278853      stranraer, n/a, united kingdom  17.0
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

In [38]:
#inspecting the columns in users_df
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


* There are records of 278858 users in this dataframe.There are 3 columns in this dataframe.
* The 'Age' column has null values

In [39]:
#Checking for duplicates in users_df
users[users['User-ID'].duplicated()].sum()

User-ID     0.0
Location    0.0
Age         0.0
dtype: float64

In [40]:
#Percentage of missing values in users_df
print(users.isnull().sum()/len(users)*100)

User-ID      0.000000
Location     0.000000
Age         39.719857
dtype: float64


* The 39.7% of values in the 'Age' column are missing/NaN values